### PR TITLE
Fixed Battle Creature Ordering

### PR DESCRIPTION
**Pull request by Ewilhan**
Fixed the messy creature order. It now follows correct ordering.

Fixes [#2837](https://bugs.vcmi.eu/view.php?id=2837).

(PS : I also removed a TODO that was wrong. Waiting moraled creatures follow the same "Slower first higher slot first" pattern as waiting creatures.)

### CODE REVIEW

File: lib/battle/BattleInfo.cpp [hunk 1]
```diff
@@ -1063,28 +1062,38 @@ bool CMP_stack::operator()(const battle::Unit * a, const battle::Unit * b)
 			int as = a->getInitiative(turn), bs = b->getInitiative(turn);
 			if(as != bs)
 				return as > bs;
-			else
-				return a->unitSlot() < b->unitSlot(); //FIXME: what about summoned stacks?
-		}
+			else if(as == bs)
```
Comment (alexvins) [inline lib/battle/BattleInfo.cpp]: Should be just `else`

File: lib/battle/BattleInfo.cpp [hunk 2]
```diff
@@ -1063,28 +1062,38 @@ bool CMP_stack::operator()(const battle::Unit * a, const battle::Unit * b)
...
-		//TODO: should be replaced with order of receiving morale!
 	case 3: //fastest last, upper slot first
 		{
 			int as = a->getInitiative(turn), bs = b->getInitiative(turn);
 			if(as != bs)
-				return as < bs;
-			else
-				return a->unitSlot() < b->unitSlot();
+				return as > bs;
+			else if(as == bs)
```
Comment (alexvins) [inline lib/battle/BattleInfo.cpp]: Should be just `else`

File: lib/battle/BattleInfo.cpp [hunk 3]
```diff
@@ -1063,28 +1062,38 @@ bool CMP_stack::operator()(const battle::Unit * a, const battle::Unit * b)
...
 		}
 	default:
 		assert(0);
 		return false;
 	}
-
 }
 
-CMP_stack::CMP_stack(int Phase, int Turn)
+CMP_stack::CMP_stack(int Phase, int Turn, int Side)
```
Comment (alexvins) [inline lib/battle/BattleInfo.cpp]: side should be `uint8_t`

File: lib/battle/BattleInfo.h [hunk 4]
```diff
@@ -144,8 +144,9 @@ class DLL_LINKAGE CMP_stack
 {
 	int phase; //rules of which phase will be used
 	int turn;
+	bool side;
```
Comment (alexvins) [inline lib/battle/BattleInfo.h]: side should be `uint8_t`

File: lib/battle/CBattleInfoCallback.cpp [hunk 5]
```diff
@@ -353,57 +353,67 @@ battle::Units CBattleInfoCallback::battleAliveUnits(ui8 side) const
...
-	if(i == all.size())
-		return nullptr;
-
-	const T * fastest = all[i], *other = nullptr;
-	int bestSpeed = fastest->getInitiative(turn);
+	const T * retCreature = nullptr;
+	size_t fastestIndex = 0;
 
-	if(fastest->unitSide() == lastMoved)
+	for(int i = 0; i < all.size(); i++)
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: `i` should be `size_t`

File: lib/battle/CBattleInfoCallback.cpp [hunk 6]
```diff
@@ -353,57 +353,67 @@ battle::Units CBattleInfoCallback::battleAliveUnits(ui8 side) const
...
+		
+		switch(phase)
 		{
-			other = all[j];
-			if(other->getInitiative(turn) != bestSpeed)
-				ret = fastest;
-			else
-				ret = other;
+		case 1: // Faster first, attacker priority, higher slot first
+			if(all[i] && (retCreature == nullptr || (curUnitSpeed > retUnitSpeed)))
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]: It could be more readable if you move `all[i]` check out of the switch

**Comment by dydzio0614** [comment]
Change names with "cur", "ret" for example retCreature to full names that are more descriptive

File: lib/battle/CBattleInfoCallback.cpp [hunk 7]
```diff
@@ -483,28 +494,37 @@ void CBattleInfoCallback::battleGetTurnOrder(std::vector<battle::Units> & out, c
...
-		if(!current)
+		const battle::Unit * current = nullptr;
+		if(phase[pi].empty())
 			pi++;
 		else
-			out.back().push_back(current);
+		{
+			current = takeOneUnit(phase[pi], actualTurn, lastMoved, pi);
+			if(!current)
+				pi++;
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]:
```suggestion
			{
				pi++;
			}
```

File: lib/battle/CBattleInfoCallback.cpp [hunk 8]
```diff
@@ -483,28 +494,37 @@ void CBattleInfoCallback::battleGetTurnOrder(std::vector<battle::Units> & out, c
...
+			else
+			{
+				out.back().push_back(current);
+				lastMoved = current->unitSide();	
+			}
+		}
 	}
 
+	if (lastMoved < 0)
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]:
```suggestion
	if(lastMoved < 0)
```

File: lib/battle/CBattleInfoCallback.cpp [hunk 9]
```diff
@@ -353,57 +353,68 @@ battle::Units CBattleInfoCallback::battleAliveUnits(ui8 side) const
...
+		int32_t currentUnitSpeed = -1;
+		int32_t returnedUnitSpeed = -1;
+		if(returnedUnit)
+			returnedUnitSpeed = returnedUnit->getInitiative(turn);
+		if(all[i])
 		{
-			if(!all[j]) continue;
-			if(all[j]->unitSide() != lastMoved || all[j]->getInitiative(turn) != bestSpeed)
+			currentUnitSpeed = all[i]->getInitiative(turn);
+			switch (phase)
```
Comment (alexvins) [inline lib/battle/CBattleInfoCallback.cpp]:
```suggestion
			switch(phase)
```

**Pull request merged**